### PR TITLE
added node def expression evaluator

### PR DESCRIPTION
**Pull request by SteRiccio**

### CODE REVIEW

File: src/record/recordExpressionEvaluator/node/identifier/identifier.ts [hunk 1]
```diff
@@ -80,12 +82,18 @@ export class RecordIdentifierEvaluator extends IdentifierEvaluator<RecordExpress
     if (!nodeDefContextUuid) {
       throw new Error(`Cannot find node with name ${propName}: context object is not a Node`)
     }
-    // node value prop
+
+    const nodeDefContext = Surveys.getNodeDefByUuid({ survey, uuid: nodeDefContextUuid })
```
Comment (ramirobg94) [inline src/record/recordExpressionEvaluator/node/identifier/identifier.ts:86]: nodeDef = getNodeDef
nodeDefContext = getNodeDefContext

**Review by ramirobg94 — CHANGES_REQUESTED** [review]

File: src/utils/index.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 export { Arrays } from './arrays'
 export { Numbers } from './numbers'
 export { Objects } from './objects'
+export { Queue } from './queue'
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/utils/index.ts:4]: uhm, this is actually a Queue object, not a module to handle queues...!

File: src/utils/queue.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+export class Queue {
+  data: any[]
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/utils/queue.ts:2]: good, right

File: src/nodeDef/nodeDefs.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { NodeDefType } from '.'
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/nodeDef/nodeDefs.ts:1]: yes, right, I did a mistake importing NodeDefType here

File: src/survey/surveys/nodeDefs.ts [hunk 5]
```diff
@@ -33,3 +33,55 @@ export const isNodeDefAncestor = (params: {
...
+    if (entitySource) {
+      children.push(...getNodeDefChildren({ survey, nodeDef: entitySource }))
+    }
+  }
+
+  children.push(
+    ...Object.values(survey.nodeDefs).filter((nodeDefCurrent) => {
+      if (nodeDefCurrent.analysis) {
+        return false
+      }
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/survey/surveys/nodeDefs.ts:74]: right, I'll add a parameter "includeAnalysis"

File: src/survey/surveys/nodeDefs.ts [hunk 6]
```diff
@@ -33,3 +33,55 @@ export const isNodeDefAncestor = (params: {
...
+
+export const getNodeDefRoot = (params: { survey: Survey }): NodeDef<NodeDefType, NodeDefProps> => {
+  const { survey } = params
+  if (!survey.nodeDefs) throw new Error('Empty node defs in survey')
+  const rootDef = Object.values(survey.nodeDefs).find((nodeDef) => !nodeDef.parentUuid)
+  if (!rootDef) throw new Error('Cannot find root def in survey')
+  return rootDef
+}
+
+export const getNodeDefSource = (params: {
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/survey/surveys/nodeDefs.ts:45]: that line should be actually:
nodeDef.virtual ? getNodeDefByUuid({ survey, uuid: nodeDef.parentUuid }) : null

File: src/utils/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramirobg94) [reply src/utils/index.ts:4]: so maybe utils it is not the right place?

File: src/survey/surveys/nodeDefs.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ramirobg94) [reply src/survey/surveys/nodeDefs.ts:45]: getNodeDefByUuid({ survey, uuid: nodeDef.parentUuid }) -> and maybe abstracted to getNodeDefParent ?

File: src/record/recordExpressionEvaluator/node/identifier/identifier.ts (hunk 1 above)
Comment (ramirobg94) [reply src/record/recordExpressionEvaluator/node/identifier/identifier.ts:86]: I dont remember :S

File: src/utils/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/utils/index.ts:4]: where can we put it? in Java for example all the classes like Map, Set, Queue, are under java.util package...!

File: src/survey/surveys/nodeDefs.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/survey/surveys/nodeDefs.ts:45]: right, we have a getNodeDefParent already

File: src/nodeDef/expressionEvaluator/node/identifier.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  }
+  return reachableNodeDefs
+}
+
+export class NodeDefIdentifierEvaluator extends IdentifierEvaluator<NodeDefExpressionContext> {
+  evaluate(expressionNode: IdentifierExpression): any {
+    try {
+      return super.evaluate(expressionNode)
+    } catch (e) {
+      // ignore it
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/nodeDef/expressionEvaluator/node/identifier.ts:71]: I will add some comments in the code to explain it...!

File: src/nodeDef/expressionEvaluator/node/identifier.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    // identifier not found
+    // identifier should be a node def or a node value property
+
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/nodeDef/expressionEvaluator/node/identifier.ts:76]: no, it was just an explanation of what will happen in the next lines

File: src/nodeDef/expressionEvaluator/node/identifier.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    const { survey, nodeDefContext } = this.context
+
+    const exprName = expressionNode.name
+
+    // identifier references a node
+    if (nodeDefContext) {
+      const reachableNodeDefs = getReachableNodeDefs({ survey, nodeDefContext })
+
+      return reachableNodeDefs.find((x: NodeDef<NodeDefType, NodeDefProps>) => x.props.name === exprName)
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/nodeDef/expressionEvaluator/node/identifier.ts:85]: right

File: src/nodeDef/expressionEvaluator/node/identifier.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  if (actualContextNode) queue.enqueue(actualContextNode)
+
+  while (!queue.isEmpty()) {
+    const entityDefCurrent = queue.dequeue()
+    const entityDefCurrentChildren = Surveys.getNodeDefChildren({ survey, nodeDef: entityDefCurrent })
+
+    reachableNodeDefs.push(entityDefCurrent, ...entityDefCurrentChildren)
+
+    // visit nodes inside single entities
+    queue.enqueueItems(entityDefCurrentChildren.filter(NodeDefs.isSingleEntity))
```
Comment (ramirobg94) [inline src/nodeDef/expressionEvaluator/node/identifier.ts:52]: instead a commend could be a function extracted

Comment (SteRiccio) [reply src/nodeDef/expressionEvaluator/node/identifier.ts:52]: ok, I'll change it ;)

File: src/nodeDef/expressionEvaluator/node/identifier.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    } catch (e) {
+      // ignore it
+    }
+
+    // identifier not found
+    // identifier should be a node def or a node value property
+
+    const { survey, nodeDefContext } = this.context
+
+    const exprName = expressionNode.name
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply src/nodeDef/expressionEvaluator/node/identifier.ts]: I've inlined that constant value

File: src/nodeDef/expressionEvaluator/node/identifier.ts (hunk 10 above)
Comment (ramirobg94) [reply src/nodeDef/expressionEvaluator/node/identifier.ts:55]: okey ;)

**Review by ramirobg94 — APPROVED** [review]

**Pull request merged**